x86-64: Implement an enhancement for byte array and char array System.arraycopy

a7ehuo opened the pull request:
Implement an enhancement for byte array and char array System.arraycopy on x86-64.

Also add JIT options that disable array copy enhancement:
- `disableArrayCopyEnhancementByteArray`: Disable array copy enhancement for 8 bit primitive array
- `disableArrayCopyEnhancementCharArray`: Disable array copy enhancement for 16 bit primitive array

a7ehuo commented:
@0xdaryl May I ask you to review this change? Thank you!

@vijaysun-omr @hzongaro fyi

a7ehuo commented:
x86-64 macOS failure is a known issue #7181
```
The following tests FAILED:
	  2 - TestJitBuilderAPIGenerator (Failed)
```

a7ehuo commented:
> Can you attach a JIT log for a method that shows the code generated for these functions?

Search `arrayCopy8BitPrimitiveEnhancementImplRoot8` for byte array or `arrayCopy16BitPrimitiveEnhancementImplRoot16` for char array to jump to the code gen section on `arraycopy`

[jitlog.arrayCopy8BitPrimitiveEnhancementImplRoot8.txt](https://github.com/eclipse/omr/files/15324938/jitlog.arrayCopy8BitPrimitiveEnhancementImplRoot8.txt)

[jitlog.arrayCopy16BitPrimitiveEnhancementImplRoot16.txt](https://github.com/eclipse/omr/files/15324939/jitlog.arrayCopy16BitPrimitiveEnhancementImplRoot16.txt)

0xdaryl commented on compiler/x/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -1767,6 +1767,367 @@ static void arrayCopy16BitPrimitive(TR::Node* node, TR::Register* dstReg, TR::Re
+   generateLabelInstruction(TR::InstOpCode::JA4, node, repMovsLabel, cg);
+
+   // 34-64 Bytes
+   generateRegMemInstruction(TR::InstOpCode::InstOpCode::VMOVDQUYmmMem, node, tmpXmmYmmReg1, generateX86MemoryReference(srcReg, sizeReg, 0, -32, cg), cg);
```
@BradleyWood : can you confirm if the use of these YMM instructions requires a check for AVX2?  This logic is only guarded with an AVX check at the moment.

BradleyWood replied:
That instruction only requires AVX.

Why not use AVX-512 to do the 64-byte load/store in a single instruction if available. Otherwise, fall back two 32-byte copies.
```
if (cg->comp()->target().cpu.supportsFeature(OMR_FEATURE_X86_AVX512F)) {
   // vmovdqu zmm0, [src]
   // vmovdqu [dest], zmm0
   generateRegMemInstruction(TR::InstOpCode::MOVDQURegMem, node, tmpXmmYmmReg2, generateX86MemoryReference(srcReg, 0, cg), cg, OMR::X86::EVEX_L512);
   generateMemRegInstruction(TR::InstOpCode::MOVDQUMemReg, node, generateX86MemoryReference(dstReg, 0, cg), tmpXmmYmmReg2, cg, OMR::X86::EVEX_L512);
} else {
   // vmovdqu ymm0, -32[src]
   // vmovdqu ymm1,   0[src]
   // vmovdqu -32[dest], ymm0
   // vmovdqu   0[dest], ymm1
   // ...
   // Use OMR::X86::VEX_l256 Encoding; assumes AVX is supported
}
```

a7ehuo replied:
Using 64-byte load/store in a single instruction is only applicable when the copy size is exactly 64 bytes. It will add an extra test when `copy size == 64` or `copy size > 64`. 

```
copy34ORMoreBytesLabel: 
   if size > 64 => repMovsLabel
   ...
repMovsLabel:
    use rep movsd for size > 64
```
==>

```
copy34ORMoreBytesLabel: 
   if size >= 64 => copy64ORMoreBytesLabel
   ...
copy64ORMoreBytesLabel:
   if size > 64 => repMovsLabel
   // 64-byte load/store in a single instruction
   ...
repMovsLabel:
    use rep movsd for size > 64
```

BradleyWood replied:
Am I missing something, because this compiles to two 32-byte loads, and two 32-byte stores = 64 bytes?
```
vmovdqu	ymm0, ymmword ptr [rsi+1*rcx-0x20]		# VMOVDQUYmmMem, SymRef [#453 -32]
vmovdqu	ymm1, ymmword ptr [rsi]		# VMOVDQUYmmMem
vmovdqu	ymmword ptr [rdi+1*rcx-0x20], ymm0		# VMOVDQUMemYmm, SymRef [#455 -32]
vmovdqu	ymmword ptr [rdi], ymm1		# VMOVDQUMemYmm
```

a7ehuo replied:
This code is used for copy size ranging from 34 to 64 bytes. It is not just for copy size as exactly 64 bytes.  For example, if we need to copy 48 bytes, it does: 

- load 32 bytes from [src + size - 32] = [src + 48 - 32] = [src + 16]  into temp1
- load 32 bytes from [src] into temp2
- store temp1 into [dst + size - 32]
- store temp2 into [dst]

```
0 1 2 ......15 16............ 31 32 33.......47
               | --------- temp1 ------------|
|----------- temp2 ------------|
```

a7ehuo replied:
I experimented using 64-byte load/store in a single instruction to copy 64 bytes. The performance is worse compared to the current implementation. As I suspect previously, I think the impact is from the extra test on the copy size. Therefore, it makes sense to  keep the current implementation for copying size 34 to 64 bytes as it is.

(1) current implementation 
(2) check if the copy size is 64 bytes, if so, using load/store in a single instruction

Copy Size | (1) ns/ops | (2) ns/ops
---|---|---
62 Bytes | 135 | 135
64 Bytes | 135 | 174
64 Bytes | 593 | 673

BradleyWood commented on compiler/x/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -1767,6 +1767,367 @@ static void arrayCopy16BitPrimitive(TR::Node* node, TR::Register* dstReg, TR::Re
+   TR::Register* tmpReg1 = cg->allocateRegister(TR_GPR);
+   TR::Register* tmpReg2 = cg->allocateRegister(TR_GPR);
+   TR::Register* tmpXmmYmmReg1 = cg->allocateRegister(TR_FPR);
+   TR::Register* tmpXmmYmmReg2 = cg->allocateRegister(TR_FPR);
```
This should be TR_VRF to use vectors

0xdaryl commented:
Generally, I think the instruction sequences are reasonable.  Please consider Brad’s suggestions where appropriate for using wider instructions on hardware that supports it.

My main feedback concerns the naming of these optimizations.  Calling them “enhancements”, while technically correct, is too generic for someone who wasn’t involved in its development up front.  There aren’t even any code comments to describe what the enhancements are, leaving others to read and understand the code to learn about them.  I suggest incorporating something like “inlineSmallArraycopiesWithoutSTOS” into the naming would make it better.  For example, “inlineSmall16BitPrimitiveArraycopiesWithoutSTOS”

Longer term, I’m not convinced that inlining this much code at each arraycopy call (over 50 instructions) is worth the simplicity and savings from calling out to a helper to do the same thing.  I’m willing to let the current inline design go through now, but I would like a helper approach evaluated for the future.  This might be fine for a method with one or two arraycopies, but if there are several arraycopies in a method (and we do see methods like that) the code footprint will add up quickly.

a7ehuo commented:
I'm currently running sanity test on the changes that address the above comments. I'll push another update once the sanity test looks good

a7ehuo commented:
@0xdaryl Comments are addressed. Ready for another review. Thank you!

- Fixed temp vector register type to `TR_VRF`
- Updated the function name and their corresponding option names to be more specific 
   - `arrayCopy8BitPrimitiveInlineSmallSizeWithoutREPMOVS`
   - `arrayCopy16BitPrimitiveInlineSmallSizeWithoutREPMOVS`
- Added comments in the beginning of the implementation to give a detailed description of the inlined code sequence

0xdaryl reviewed (commented):
Thanks for adding the comments describing the operation.  They look good!

I missed publishing a couple of other comments in my earlier review for some reason.  They are minor changes only.

I realize the test farm isn't in great shape right now.  Can you confirm you tested this code internally on both 64-bit and 32-bit?

0xdaryl commented:
@BradleyWood : please approve if your comments are addressed

a7ehuo commented:
> Can you confirm you tested this code internally on both 64-bit and 32-bit?

I tested only on 64-bit machines since I thought the enhancement was guarded with checking if target is 64-bit. I'll test on 32-bit as well after addressing the latest comments. 

```  
bool enable8BitPrimitiveArrayCopyInlineSmallSizeWithoutREPMOVS = ((elementSize == 1) &&
                                                                     !disableEnhancement &&
                                                                     cg->comp()->target().cpu.supportsAVX() &&
                                                                     cg->comp()->target().is64Bit()) ? true : false;
```

```
bool enable16BitPrimitiveArrayCopyInlineSmallSizeWithoutREPMOVS = (!disableEnhancement &&
                                                                 cg->comp()->target().cpu.supportsAVX() &&
                                                                 cg->comp()->target().is64Bit()) ? true : false;
```

a7ehuo commented on compiler/x/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -1767,6 +1767,367 @@ static void arrayCopy16BitPrimitive(TR::Node* node, TR::Register* dstReg, TR::Re
+   generateRegImmInstruction(TR::InstOpCode::CMPRegImm4(), node, sizeReg, 2, cg);
+   generateLabelInstruction(TR::InstOpCode::JA4, node, copy3ORMoreBytesLabel, cg);
+
+   generateRegImmInstruction(TR::InstOpCode::CMPRegImm4(), node, sizeReg, 0, cg);
```
Addressed in [a81d5c0](https://github.com/eclipse/omr/compare/9065569a6cc25dc6c3757e54a807e6fbf854188d..a81d5c0bb861ee5cc3508f8cb9df9c7f281c3a65)

a7ehuo commented on compiler/x/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -1767,6 +1767,367 @@ static void arrayCopy16BitPrimitive(TR::Node* node, TR::Register* dstReg, TR::Re
+   generateRegImmInstruction(TR::InstOpCode::CMPRegImm4(), node, sizeReg, 2, cg);
+   generateLabelInstruction(TR::InstOpCode::JA4, node, copy4ORMoreBytesLabel, cg);
+
+   generateRegImmInstruction(TR::InstOpCode::CMPRegImm4(), node, sizeReg, 0, cg);
```
Addressed in [a81d5c0](https://github.com/eclipse/omr/compare/9065569a6cc25dc6c3757e54a807e6fbf854188d..a81d5c0bb861ee5cc3508f8cb9df9c7f281c3a65)

a7ehuo commented:
@0xdaryl Addressed all review comments. Ready for another review. Thank you!

- Internal Jenkins test (ID 22283): JDK 8 `sanity.functional,sanity.system,sanity.openjdk,extended.system,extended.functional` with `x86-32_windows`, all passed. 
- Internal Jenkins test (ID 22275): JDK11 and JDK 17  `sanity.functional,sanity.system,sanity.openjdk,extended.system,extended.functional` with `x86-64_linux` : Seen some failures and the failures are not related to this change

0xdaryl commented:
@vijaysun-omr : FYI

vijaysun-omr commented:
Since we want to keep all codegens aware of key changes in one, I thought I'd mention @zl-wang @r30shah and @knn-k here so that they can consider whatever feels relevant to their platform
